relay-event-schema/src/protocol/span/convert.rs

```diff
@@ -0,0 +1,254 @@
+                    )*
+                    $(
+                        $(
+                            $foo: event.context::<$ContextType>()
+                                .map_or(None, |ctx|ctx.$context_field.value().cloned()).into(),
+                            $(
+                                $(
+                                    $span_field_from_context: event.context::<$ContextType>()
+                                    .map_or(None, |ctx|ctx.$context_field.value().cloned()).into(),
+                                )+
+                            )?
+                        )*
+                    )*
+                    $(
+                        $fixed_span_field: $fixed_span_value.into(),
+                    )*
+                    ..Default::default()
+                }
+            }
+        }
+
+        #[allow(clippy::needless_update)]
+        impl From<&Span> for Event {
+            fn from(span: &Span) -> Self {
+                Self {
+                    $(
+                        $event_field: span.$span_field.clone(),
+                    )*
+                    $(
+                        $fixed_event_field: $fixed_event_value.into(),
+                    )*
+                    contexts: Annotated::new(
+                        Contexts(
+                            BTreeMap::from([
+                                $(
+                                    (<$ContextType as DefaultContext>::default_key().into(), ContextInner($ContextType {
+                                        $(
+                                            $context_field: span.$foo.clone(),
+                                        )*
+                                        ..Default::default()
+                                    }.into_context()).into()),
+                                )*
+                            ]),
+                        )
+                    ),
+                    ..Default::default()
+                }
+            }
+        }
+    };
+}
+
+// This macro call implements a bidirectional mapping between transaction event and segment spans,
+// allowing users to call both `Event::from(&span)` and `Span::from(&event)`.
+map_fields!(
+    top-level:
+        span._metrics_summary <=> event._metrics_summary,
+        span.description <=> event.transaction,
+        span.measurements <=> event.measurements,
+        span.platform <=> event.platform,
+        span.received <=> event.received,
+        span.start_timestamp <=> event.start_timestamp,
+        span.timestamp <=> event.timestamp
+    ;
+    contexts:
+        TraceContext:
+            span.exclusive_time <=> context.exclusive_time,
+            span.op <=> context.op,
+            span.parent_span_id <=> context.parent_span_id,
+            // A transaction corresponds to a segment span, so span_id and segment_id have the same value:
+            span.span_id, span.segment_id <=> context.span_id,
```

Comment:
that is interesting syntax, should we just split it into two different lines?

Reply:
What's the value in `event.context.span_id` if the `span.span_id` and `span.segment_id` are different?

Reply:
The first one. I improved the variable names to clarify:

https://github.com/getsentry/relay/blob/1809a0f7e8cb5fc15febb6268879a995bf6cb3d2/relay-event-schema/src/protocol/span/convert.rs#L94-L97